### ISSUE TITLE
CompileManager::compile() is single-threaded and sleeps too much

### ISSUE THREAD

**Issue by timoore** [vsg-dev/VulkanSceneGraph#783]
The bug is really "data compilation is really slow in my application." in vsgCs it takes on average 100ms to "compile" a 256x479 (ish) texture. This is a continuation of #782 with an eye to writing down the various performance problems with CompileManager::compile() and inching towards a solution.

In VSG usage, "compile" means "create Vulkan objects and upload to the GPU everything needed to render a scenegraph." The main entry point to compilation is through the CompileManager::compile() function which is called when files are loaded at the beginning of an application and when new data is added via the DatabasePager. The problems described here were encountered in vsgCs which doesn't use the DatabasePager, but the issues are the same.

The data uploaded by compilation, which are usually textures and geometry for rendering, must often be further processed by commands that are dispatched to Vulkan, e.g. to copy textures to fast GPU memory or to create mipmaps. A complication is that this compilation often must be repeated for different usages of the graphics data, due to the way in which Vulkan manages graphical state through "pipeline objects." In VSG a Context object exists for each of these usages; broadly speaking, a context is created for each View and rendering method. At the lowest levels of the VSG scenegraph, different Vulkan pipeline objects are stored for each context. New contexts may be added as a program progresses. compile() does not (intentionally!) touch the Viewer's scene graph; the database pager or application does that in the main thread, using results of compile().

This compilation should be very parallelizable, thanks to the design of Vulkan. The slowest parts -- compilation of shaders from source and pipeline creation -- can be run in parallel with each other and rendering. It's a bit tricky to submit commands to a single Vulkan queue from several threads, but VSG already deals with that. The problems come with synchronization among the several actors:

* The Vulkan spec is fairly non-commital about whether operations on a queue can block or not. VSG queue operations might end up sleeping inside the mutex that protects the main graphics queue, blocking any other queue operations.
* Vulkan temporary memory used by compile() needs to be cleaned up after it is used. This is currently done by waiting -- on the CPU -- on a fence after each compile operation and then doing the cleanup. In the current source code, this prevents any other compilation operation from starting.
* Vulkan rendering commands can't use uploaded data until they are ready. The waiting CPU-side on a fence does take care of this, but using a semaphore during command submission would allow work to proceed on the CPU.
* The VSG low-level data structures can't be changed while in use.  There are several ways this might happen, but the current serial compile basically protects against this.

The last point is complicated to get right because it's common to have shared structure in the scene graph. Any kind of model library is naturally going to share structure in order to conserve resources. compile() generally won't recompile parts of the scene graph, but two parallel compile processes might get in a race and end up stomping on each other.

So, all of this is to set the stage for my complaint above. vsgCs does compilation in serveral threads and it's possible that serializing calls to CompileManager:::compile() would help the total time, because much of this time is waiting for a compile traversal (compilation task) to become available. Even so, there is a huge amount of latency in compilation for a process that ought to be 1) fast and 2) quite parallel. I can see getting there by attacking the synchronization issues above (maybe these should be turned into their own issues):

* Compilation cleanup and sync with Vulkan could be elegantly handled with a timeline semaphore that would insure, when rendering commands are submitted, that all previous compile operations have completed. Cleanup could be deferred to that same submission, in the recording traversal thread.
* A separate graphics queue and / or transfer queue for compilation might be helpful.
* Open Scene Graph has a lot of ceremony associated with access to the scene graph and dynamic data, and perhaps a similar notion of critical sections where it is OK to change the low-level scene graph is in order. Another approach would be to synchronize the creation of new contexts and pipelines, which is one of the only places that a race condition caused by VSG could really occur. Rules about sharing scene graph structure need to be formalized.

Note: I'm not considering any issues brought about by having multiple devices or adding them at run time.

**Comment by robertosfield** [comment]
Unfortunately I'm really busy with client work, all VSG related, but it does mean I don't have much time to look at issue like this right away.

I think there are a several issues all bound up together which will need addressing either with a refactor refinement of the core VSG w.r.t compiling shaders, creating Vulakn objects and transfer data to the GPU.  For different usage cases one of these refactor/refinements may be sufficient to resolve the issue.

Part of the solution may also be in how you create the scene graph as it may be that more work can be moved out of the compile traversal - such as recompiling ShaderSet variants.   The vsgshaderset example/utility illustrates how this can be done.

It may also be that a particular Vulkan driver/GPU combination struggles with particular types of data and avoiding this may alleviate the slow handling of that data type.  For instance RGB image data can't be handled directly for texturing, it has to first be converted to RGBA which is slow when done on the CPU.  The vsgdynamictexture_cs example illustrates conversion of RGB to RGBA as one alternative approach.  Doing the conversion on the CPU, within the database paging thread prior to the compile traversal is another way of avoiding the compile traversal from being slowed down.

**Comment by timoore** [comment]
> Unfortunately I'm really busy with client work, all VSG related, but it does mean I don't have much time to look at issue like this right away.
> 
No problem. It's not a critical bug, doesn't affect frame rate etc., but it does introduce unnecessary latency in apps that do database paging.

> I think there are a several issues all bound up together which will need addressing either with a refactor refinement of the core VSG w.r.t compiling shaders, creating Vulakn objects and transfer data to the GPU. For different usage cases one of these refactor/refinements may be sufficient to resolve the issue.

Yup.
> 
> Part of the solution may also be in how you create the scene graph as it may be that more work can be moved out of the compile traversal - such as recompiling ShaderSet variants. The vsgshaderset example/utility illustrates how this can be done.

That particular problem is not a performance issue; I've done a lot of profiling and tracing of the paging behavior. vsgCs compiles doesn't adjust the  shaders it uses much during the lifetime of the app, but the compiler latency is always there. There probably is a data corruption problem right now in ShaderCompiler, but that's just one of the things preventing the compile traversal from running parallel.
> 

> It may also be that a particular Vulkan driver/GPU combination struggles with particular types of data and avoiding this may alleviate the slow handling of that data type. For instance RGB image data can't be handled directly for texturing, it has to first be converted to RGBA which is slow when done on the CPU. The vsgdynamictexture_cs example illustrates conversion of RGB to RGBA as one alternative approach. Doing the conversion on the CPU, within the database paging thread prior to the compile traversal is another way of avoiding the compile traversal from being slowed down.

Those are all potential issues when the code is running, but the main problem is the code that is not running.

**Comment by robertosfield** [comment]
You mention "average 100ms to "compile" a 256x479 (ish) texture."  

What pixel format does the data have?  Does it require mipmap generation.

Does a similar cost happen when you load data which isn't from vsgCs?

**Comment by timoore** [comment]
> You mention "average 100ms to "compile" a 256x479 (ish) texture."
> 
> What pixel format does the data have? Does it require mipmap generation.
It's all R8G8B8_SRGB. I do the conversion from that to RGBA myself. I do rely on VSG to do the mipmap generation cuz that's the right way to do SRGB mipmaps. My prep averages 1.2ms, which isn't great; this does not count decoding the JPEG source file. have an idea for optimizing the RGB->RGBA conversion and will look into your compute shader suggestion. Also, I think there might be issues with VSG's  memory allocator and multithreading, but I don't have much to say about that yet.

I will see if not doing GPU mipmap generation makes a difference.

> 
> Does a similar cost happen when you load data which isn't from vsgCs?
I don't know.

**Comment by timoore** [comment]
See #786.

**Comment by robertosfield** [comment]
Is this topic resolved enough to close now?

**Comment by timoore** [comment]
Sure; it might become a discussion again.
